pkg/controller/datavolume/upload-controller.go

```diff
@@ -93,50 +92,47 @@ func NewUploadController(
 	return datavolumeController, nil
 }
 
-func (r UploadReconciler) updateAnnotations(dataVolume *cdiv1.DataVolume, annotations map[string]string) error {
+func (r UploadReconciler) updateAnnotations(dataVolume *cdiv1.DataVolume, pvc *corev1.PersistentVolumeClaim) error {
 	if dataVolume.Spec.Source.Upload == nil {
 		return errors.Errorf("no source set for upload datavolume")
 	}
-	annotations[cc.AnnUploadRequest] = ""
+	pvc.Annotations[cc.AnnUploadRequest] = ""
 	return nil
 }
 
-func (r UploadReconciler) updateStatus(log logr.Logger, syncRes dataVolumeSyncResult, syncErr error) (reconcile.Result, error) {
-	if syncErr != nil {
-		return reconcile.Result{}, syncErr
-	}
-	if syncRes.res != nil {
-		return *syncRes.res, nil
-	}
-	//FIXME: pass syncRes instead of args
-	if err := r.reconcileUpload(log, syncRes.dv, &syncRes.pvc, syncRes.pvcSpec); err != nil {
-		return reconcile.Result{}, err
-	}
-	return r.reconcileDataVolumeStatus(syncRes.dv, syncRes.pvc, nil, r.updateStatusPhase)
+// Reconcile loop for the upload data volumes
+func (r UploadReconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	log := r.log.WithValues("DataVolume", req.NamespacedName)
+	return r.updateStatus(r.sync(log, req))
 }
 
-func (r *UploadReconciler) reconcileUpload(log logr.Logger, dv *cdiv1.DataVolume, pvc **corev1.PersistentVolumeClaim, pvcSpec *v1.PersistentVolumeClaimSpec) error {
-	if _, dvPrePopulated := dv.Annotations[cc.AnnPrePopulated]; dvPrePopulated {
-		return nil
-	}
-	if *pvc != nil {
-		return nil
+func (r UploadReconciler) sync(log logr.Logger, req reconcile.Request) (dataVolumeSyncResult, error) {
+	var syncRes dataVolumeSyncResult
+	syncErr := r.syncCommon(log, req, &syncRes, nil, nil)
+	if syncErr == nil && syncRes.result == nil && syncRes.pvc == nil {
+		if _, dvPrePopulated := syncRes.dv.Annotations[cc.AnnPrePopulated]; !dvPrePopulated {
+			syncRes.pvc, syncErr = r.createPvcForDatavolume(syncRes.dv, syncRes.pvcSpec, r.updateAnnotations)
+		}
 	}
+	return syncRes, syncErr
+}
 
-	newPvc, err := r.createPvcForDatavolume(dv, pvcSpec, nil)
-	if err != nil {
-		if cc.ErrQuotaExceeded(err) {
-			r.updateDataVolumeStatusPhaseWithEvent(cdiv1.Pending, dv, nil, nil,
+func (r UploadReconciler) updateStatus(syncRes dataVolumeSyncResult, syncErr error) (reconcile.Result, error) {
+	if syncErr != nil {
+		if cc.ErrQuotaExceeded(syncErr) {
+			r.updateDataVolumeStatusPhaseWithEvent(cdiv1.Pending, syncRes.dv, nil, nil,
 				Event{
 					eventType: corev1.EventTypeWarning,
 					reason:    cc.ErrExceededQuota,
-					message:   err.Error(),
+					message:   syncErr.Error(),
 				})
 		}
-		return err
+		return reconcile.Result{}, syncErr
```

Comment:
Why is updateStatusCommon not called?  The main idea of this pattern is that even if there is an error during sync, the status is checked/updated if anything changes.  Do you know for sure that the status has not changed from whatever it was last?

Reply:
fixed